### PR TITLE
feat: Allow checking statuscodes of batch response without parsing step body

### LINKED ISSUE #621: BatchResponseContent and BatchResponseContentCollection should have validate success
Batching should make your application a lot faster, but the SDK has no way to check if all modifications are successful without checking each individual request.

I'm also the creator of #613 so I know a thing or two on batching. This new proposal needs addition input, because I'm not sure on the solution either.

Please provide the following (and please check them off the list with [x]) before submitting this issue:
- [x] Expected behavior. Please provide **links to the specific [Microsoft Graph documentation](https://developer.microsoft.com/en-us/graph/docs/concepts/overview)** you used to determine the expected behavior. 
- [x] Actual behavior. Provide error codes, stack information, and a [Fiddler](http://www.telerik.com/fiddler) capture of the request and response (please remove personally identifiable information before posting).
- [x] Steps to reproduce the behavior. Include your code, IDE versions, client library versions, and any other information that might be helpful to understand your scenario.

### Expected behavior

If you added 15 items to a batch and it's executed I would expect there to be a way to check if all individual requests have a success response code (`200`, `202` or `204`).

Something like:

```csharp
var batchCollection = new BatchRequestContentCollection(graphClient);
foreach (var task in tasks)
{
    await batchCollection.AddBatchRequestStepAsync(graphClient.Me.Todo.Lists[listId].Tasks.ToPostRequestInformation(task));
}
var result = await graphClient.Batch.PostAsync(batchCollection);
var success = await result.ValidateSuccess();
```

And maybe even return the failed requests somehow for retry.

### Actual behavior

`var success = await result.ValidateSuccess();` is not available.

### Steps to reproduce the behavior

Run [this program](https://github.com/svrooij/msgraph-sdk-dotnet-batching/tree/main/sample/BatchClient) and see the batch execute successfully, while individual requests not all succeed (16 of the 20 requests are blocked with a `429` error code).

### PR DESCRIPTION
**Pull request by svrooij**
Fixes #621 

### Changes proposed in this pull request
- Expose `GetResponsesStatusCodesAsync()` for **BatchResponseContentCollection** and **BatchResponseContent** that allows checking the status codes without creating **HttpResponseMessage** for each step (that has to be disposed)
- Expose `BatchRequestSteps` for **BatchRequestContentCollection** so you can possibly fetch the failed steps to retry, for instance in case of some `429` results in the batch.
- extra constructor for `BatchRequestContentCollection` including the number of requests that may be combined. (At the moment you cannot combine more then 4 requests, when creating TodoTasks, all above 4 fail with 429.) This allows the user to use batching without having to manage smaller batches himself.

### Other links
- https://svrooij.io/2023/03/03/batching-in-microsoft-graph/
- Demo code https://github.com/svrooij/msgraph-sdk-dotnet-batching/tree/main/sample/BatchClient
- Video demonstrating batching https://youtu.be/u9buNZxbSNk

@maisarissi I've made batching even better to also support getting the failing steps.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/pull/626)

### CODE REVIEW

**Review by andrueastman — COMMENTED** [review]
Thanks again for this.

File: src/Microsoft.Graph.Core/Requests/Content/BatchRequestContentCollection.cs [hunk 1]
```diff
@@ -13,16 +14,32 @@ public class BatchRequestContentCollection
...
+        /// </summary>
+        /// <param name="baseClient">The <see cref="IBaseClient"/> for making requests</param>
+        /// <param name="batchRequestLimit">Number of requests that may be placed in a single batch</param>
+        public BatchRequestContentCollection(IBaseClient baseClient, int batchRequestLimit)
+        {
+            if (batchRequestLimit < 2 || batchRequestLimit > CoreConstants.BatchRequest.MaxNumberOfRequests)
+            {
+                throw new ArgumentOutOfRangeException(nameof(batchRequestLimit));
+            }
             this.baseClient = baseClient;
```
[the remark this answers is not in the document]
Comment (svrooij) [reply src/Microsoft.Graph.Core/Requests/Content/BatchRequestContentCollection.cs:45]: Sure, great idea

**Comment by maisarissi** [comment]
Thanks @svrooij again, for the proposed enhancements!

**Comment by svrooij** [comment]
@maisarissi does Microsoft also have a digital badge for "brilliant ideas"? And the mean reason for this latest PR was that we are faced with a limit of 4 requests when trying to modify TodoTasks. Where should I post about that issue?

**Comment by maisarissi** [comment]
Hey @svrooij . I'm not aware of such badge, but we are working on a new program related to SDKs :) Once the program is ready, we will communicate how things are going to work.
For the second question around TodoTasks, this is an API limitation, so I would say the best place to ask questions/features is the Microsoft Q&A: https://developer.microsoft.com/en-us/graph/support

**Pull request merged**